frame/dex-router/src/lib.rs

```diff
@@ -0,0 +1,272 @@
+			updated_route: Vec<DexRouteNode<T::PoolId>>,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+
+	impl<T: Config> Pallet<T> {
+		fn do_update_route(
+			who: &T::AccountId,
+			asset_pair: CurrencyPair<T::AssetId>,
+			route: BoundedVec<DexRouteNode<T::PoolId>, T::MaxHopsInRoute>,
+		) -> Result<(), DispatchError> {
+			let k1 = asset_pair.base;
+			let k2 = asset_pair.quote;
+			for r in route.as_slice() {
+				match r {
+					DexRouteNode::Curve(pool_id) => {
+						ensure!(
+							T::StableSwapDex::pool_exists(*pool_id),
+							Error::<T>::PoolDoesNotExist
+						)
+					},
+					DexRouteNode::Uniswap(pool_id) => {
+						ensure!(
+							T::ConstantProductDex::pool_exists(*pool_id),
+							Error::<T>::PoolDoesNotExist
+						)
+					},
+				}
+			}
+			let existing_route = DexRoutes::<T>::get(k1, k2);
+
+			DexRoutes::<T>::insert(k1, k2, DexRoute::Direct(route.clone()));
+			let event = match existing_route {
+				Some(DexRoute::Direct(old_route)) => Event::RouteUpdated {
+					who: who.clone(),
+					x_asset_id: k1,
+					y_asset_id: k2,
+					old_route: old_route.into_inner(),
+					updated_route: route.to_vec(),
+				},
+				None => Event::RouteAdded {
+					who: who.clone(),
+					x_asset_id: k1,
+					y_asset_id: k2,
+					route: route.to_vec(),
+				},
+			};
+			Self::deposit_event(event);
+			Ok(())
+		}
+
+		fn do_delete_route(
+			who: &T::AccountId,
+			asset_pair: CurrencyPair<T::AssetId>,
+		) -> Result<(), DispatchError> {
+			let k1 = asset_pair.base;
+			let k2 = asset_pair.quote;
+			if let Some(DexRoute::Direct(deleted_route)) = DexRoutes::<T>::take(k1, k2) {
+				Self::deposit_event(Event::RouteDeleted {
+					who: who.clone(),
+					x_asset_id: k1,
+					y_asset_id: k2,
+					route: deleted_route.into_inner(),
+				});
+			}
+
+			Ok(())
+		}
+	}
+
+	impl<T: Config> DexRouter<T::AccountId, T::AssetId, T::PoolId, T::Balance, T::MaxHopsInRoute>
+		for Pallet<T>
+	{
+		fn update_route(
+			who: &T::AccountId,
+			asset_pair: CurrencyPair<T::AssetId>,
+			route: Option<BoundedVec<DexRouteNode<T::PoolId>, T::MaxHopsInRoute>>,
+		) -> Result<(), DispatchError> {
+			match route {
+				Some(bounded_route) => Self::do_update_route(who, asset_pair, bounded_route)?,
+				None => Self::do_delete_route(who, asset_pair)?,
+			}
+			Ok(())
+		}
+
+		fn get_route(asset_pair: CurrencyPair<T::AssetId>) -> Option<Vec<DexRouteNode<T::PoolId>>> {
+			DexRoutes::<T>::get(asset_pair.base, asset_pair.quote)
+				.map(|DexRoute::Direct(route)| route.into_inner())
+		}
+
+		fn exchange(
+			who: &T::AccountId,
+			asset_pair: CurrencyPair<T::AssetId>,
+			dx: T::Balance,
+		) -> Result<T::Balance, DispatchError> {
+			let route = Self::get_route(asset_pair).ok_or(Error::<T>::NoRouteFound)?;
+			let mut dx_t = dx;
+			let mut dy_t = T::Balance::zero();
+			for route_node in &route {
+				match route_node {
+					DexRouteNode::Curve(pool_id) => {
+						dy_t = T::StableSwapDex::exchange(
+							who,
```

Comment:
let x = match dex { a => T::StableSwapDex::exchange, b => DexRouteNode::Uniswap}.

x(...all parameters here..).

Reply:
are you ok doing this in next PR?

Reply:
if yes, add TODO here

Reply:
I tried doing but I see compilation errors. Please do the changes after this PR is done.